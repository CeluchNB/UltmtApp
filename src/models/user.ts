import { Realm } from '@realm/react'

export const DisplayUserSchema: Realm.ObjectSchema = {
    name: 'DisplayUser',
    embedded: true,
    properties: {
        _id: 'string',
        firstName: 'string',
        lastName: 'string',
        username: 'string',
        localGuest: 'bool?',
    },
}
