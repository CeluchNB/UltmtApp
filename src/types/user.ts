import { DisplayTeam } from './team'

export interface CreateUserData {
    firstName: string
    lastName: string
    username: string
    email: string
    password: string
}

export interface DisplayUser {
    _id: string
    firstName: string
    lastName: string
    username: string
}

export interface User extends DisplayUser {
    email: string
    requests: string[]
    playerTeams: DisplayTeam[]
    managerTeams: DisplayTeam[]
    archiveTeams: DisplayTeam[]
    openToRequests: boolean
    private: boolean
    guest: boolean
}

export interface GuestUser {
    _id?: string
    firstName: string
    lastName: string
    username?: string
}
