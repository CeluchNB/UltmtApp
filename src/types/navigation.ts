import { DisplayTeam } from './team'
import { DisplayUser } from './user'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import { RequestType } from './request'

export enum SecureEditField {
    EMAIL,
    PASSWORD,
}

export type TopLevelParamList = {
    Tabs: undefined
    SettingsScreen: undefined
    SecureEditScreen: { title: string; value: string; field: SecureEditField }
}

export type AccountStackParamList = {
    Login: undefined
    Profile: undefined
    CreateAccount: undefined
    CreateTeam: undefined
    ForgotPasswordScreen: undefined
    ManageTeams: undefined
    RequestTeam: undefined
    ManagedTeamDetails: { id: string; place: string; name: string }
    PublicTeamDetails: {
        id: string
        place: string
        name: string
        archive?: boolean
    }
    PublicUserDetails: { user: DisplayUser }
    RequestUser: { type: RequestType }
    ResetPasswordScreen: undefined
    RolloverTeam: undefined
    UserRequestsScreen: undefined
    TeamRequestsScreen: undefined
    JoinByCodeScreen: undefined
}

export type GameStackParamList = {
    GameHome: undefined
    GameSearch: { live: string }
    CreateGame: { teamOne: DisplayTeam }
    SelectOpponent: undefined
}

export type AllScreenProps = NativeStackScreenProps<
    AccountStackParamList & TopLevelParamList & GameStackParamList,
    'Tabs'
>

export type TopLevelProps = NativeStackScreenProps<
    TopLevelParamList,
    'SettingsScreen'
>

export type SecureEditProps = NativeStackScreenProps<
    TopLevelParamList,
    'SecureEditScreen'
>

export type Props = NativeStackScreenProps<AccountStackParamList, 'Login'>

export type CreateTeamProps = NativeStackScreenProps<
    AccountStackParamList,
    'CreateTeam'
>
export type ManagedTeamDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'ManagedTeamDetails'
>

export type PublicTeamDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'PublicTeamDetails'
>

export type PublicUserDetailsProps = NativeStackScreenProps<
    AccountStackParamList,
    'PublicUserDetails'
>

export type RequestUserProps = NativeStackScreenProps<
    AccountStackParamList,
    'RequestUser'
>

export type GameSearchProps = NativeStackScreenProps<
    GameStackParamList,
    'GameSearch'
>

export type CreateGameProps = NativeStackScreenProps<
    GameStackParamList,
    'CreateGame'
>
