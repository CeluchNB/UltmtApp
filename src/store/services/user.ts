import { ApiResponse } from '../../types/services'
import { CreateUserData } from '../../types/user'

export const login = async (
    username: string,
    password: string,
): Promise<ApiResponse> => {
    const response = await fetch('https://ultmt-dev.herokuapp.com/user/login', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: '*/*',
        },
        body: JSON.stringify({
            email: username,
            password: password,
        }),
    })

    if (response.ok) {
        return { data: await response.json() }
    } else {
        throw new Error(await response.text())
    }
}

export const fetchProfile = async (token: string): Promise<ApiResponse> => {
    const response = await fetch('https://ultmt-dev.herokuapp.com/user/me', {
        method: 'GET',
        headers: {
            Authorization: `Bearer ${token}`,
        },
    })

    if (response.ok) {
        return { data: await response.json() }
    } else {
        throw new Error(await response.text())
    }
}

export const createAccount = async (
    profileData: CreateUserData,
): Promise<ApiResponse> => {
    console.log('making create account call')
    const response = await fetch('https://ultmt-dev.herokuapp.com/user', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: '*/*',
        },
        body: JSON.stringify({ ...profileData }),
    })

    if (response.ok) {
        return { data: await response.json() }
    } else {
        throw new Error(await response.text())
    }
}
