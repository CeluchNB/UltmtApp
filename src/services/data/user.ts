import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import EncryptedStorage from 'react-native-encrypted-storage'
import jwt_decode from 'jwt-decode'
import { saveTeams as localSaveTeams } from '../local/team'
import { throwApiError } from '../../utils/service-utils'
import { withToken } from './auth'
import { CreateUserData, DisplayUser, User } from '../../types/user'
import {
    changeEmail as networkChangeEmail,
    changeName as networkChangeName,
    changePassword as networkChangePassword,
    createAccount as networkCreateAccount,
    deleteAccount as networkDeleteAccount,
    fetchProfile as networkFetchProfile,
    getPublicUser as networkGetPublicUser,
    joinTeamByCode as networkJoinTeamByCode,
    leaveManagerRole as networkLeaveManagerRole,
    leaveTeam as networkLeaveTeam,
    requestPasswordRecovery as networkRequestPasswordRecovery,
    resetPassword as networkResetPassword,
    searchUsers as networkSearchUsers,
    setOpenToRequests as networkSetOpenToRequests,
    setPrivate as networkSetPrivate,
} from '../network/user'

/**
 * Method to get a user's profile by their jwt
 * @returns a user object
 * @throws error if backend returns an error
 */
export const fetchProfile = async (): Promise<User> => {
    try {
        console.log('calling fetch profile')
        const response = await withToken(networkFetchProfile)
        const { user, fullManagerTeams } = response.data
        await localSaveTeams(fullManagerTeams)
        return user
    } catch (error: any) {
        return throwApiError(error, Constants.GENERIC_FETCH_PROFILE_ERROR)
    }
}

/**
 * Method to create
 * @param profileData data to create account
 * @returns user data and a jwt token
 * @throws error if backend returns an error
 */
export const createAccount = async (
    profileData: CreateUserData,
): Promise<{ user: User }> => {
    try {
        const response = await networkCreateAccount(profileData)
        const { user, tokens } = response.data
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
        return { user }
    } catch (error: any) {
        return throwApiError(error, Constants.GENERIC_CREATE_ACCOUNT_ERROR)
    }
}

/**
 * Method to search for users
 * @param term search term
 * @returns list of users
 * @throws error if backend returns an error
 */
export const searchUsers = async (term: string): Promise<DisplayUser[]> => {
    try {
        if (term.length < 3) {
            throw new ApiError('Not enough characters to search')
        }
        const response = await networkSearchUsers(term)
        return response.data.users
    } catch (error) {
        throw throwApiError(error, Constants.SEARCH_ERROR)
    }
}

/**
 * Method for user to leave team
 * @param teamId id of team to leave
 * @returns updated user profile
 * @throws error if backend returns an error
 */
export const leaveTeam = async (teamId: string): Promise<User> => {
    try {
        const response = await withToken(networkLeaveTeam, teamId)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.LEAVE_TEAM_ERROR)
    }
}

/**
 * Method to get a user's public details. No authentication needed to make this call.
 * @param id id of user
 * @returns user object
 * @throws error if backend returns an error
 */
export const getPublicUser = async (id: string): Promise<User> => {
    try {
        const response = await networkGetPublicUser(id)
        const user = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.GENERIC_FETCH_PROFILE_ERROR)
    }
}

/**
 * Method for user to remove himself as manager of a team
 * @param teamId team id to leave
 * @returns updated user
 */
export const leaveManagerRole = async (teamId: string): Promise<User> => {
    try {
        const response = await withToken(networkLeaveManagerRole, teamId)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method for user to request a code to recover their password
 * @param email email associated with account
 * @returns void
 */
export const requestPasswordRecovery = async (email: string): Promise<void> => {
    try {
        await networkRequestPasswordRecovery(email)
    } catch (error) {
        return throwApiError(error, Constants.UNABLE_TO_EMAIL)
    }
}

/**
 * Method for user to reset their password
 * @param passcode 6 digit passcode user can redeem
 * @param newPassword new password
 * @returns new jwt and user profile
 */
export const resetPassword = async (
    passcode: string,
    newPassword: string,
): Promise<{ user: User }> => {
    try {
        const response = await networkResetPassword(passcode, newPassword)
        const { tokens, user } = response.data
        const { access, refresh } = tokens
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
        return { user }
    } catch (error) {
        return throwApiError(error, Constants.RESET_PASSWORD_ERROR)
    }
}

/**
 * Method to change user's ability to receive requests
 * @param open boolean if user wishes to receive requests
 * @returns updated user profile
 */
export const setOpenToRequests = async (open: boolean): Promise<User> => {
    try {
        const response = await withToken(networkSetOpenToRequests, open)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.TOGGLE_ROSTER_STATUS_ERROR)
    }
}

/**
 * Method to update a user's first and last name
 * @param firstName first name of user
 * @param lastName last name of user
 * @returns updated user value
 */
export const changeName = async (
    firstName: string,
    lastName: string,
): Promise<User> => {
    try {
        const response = await withToken(networkChangeName, firstName, lastName)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_USER_ERROR)
    }
}

/**
 * Method to set a user's private status
 * @param privateAccount user's private status
 * @returns updated user value
 */
export const setPrivate = async (privateAccount: boolean): Promise<User> => {
    try {
        const response = await withToken(networkSetPrivate, privateAccount)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_USER_ERROR)
    }
}

/**
 * Method to change a user's email
 * @param email current email to fulfill email/password login requirement
 * @param password current password
 * @param newEmail new email
 * @returns updated user value
 */
export const changeEmail = async (
    email: string,
    password: string,
    newEmail: string,
): Promise<User> => {
    try {
        const response = await networkChangeEmail(email, password, newEmail)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_USER_ERROR)
    }
}

/**
 * Method to change a user's password
 * @param email current email to fulfill email/password login requirement
 * @param password current password
 * @param newPassword updated password
 * @returns updated user value
 */
export const changePassword = async (
    email: string,
    password: string,
    newPassword: string,
): Promise<User> => {
    try {
        const response = await networkChangePassword(
            email,
            password,
            newPassword,
        )
        const {
            user,
            tokens: { access, refresh },
        } = response.data
        await EncryptedStorage.setItem('access_token', access)
        await EncryptedStorage.setItem('refresh_token', refresh)
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_USER_ERROR)
    }
}

/**
 * Method to delete a user's account
 * @returns nothing
 */
export const deleteAccount = async (): Promise<void> => {
    try {
        await withToken(networkDeleteAccount)
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
    } catch (error) {
        await EncryptedStorage.removeItem('access_token')
        await EncryptedStorage.removeItem('refresh_token')
        return throwApiError(error, Constants.DELETE_USER_ERROR)
    }
}

/**
 * Method to join a team by a 6 digit bulk code
 * @param code 6 digit bulk team join code
 * @returns updated user profile
 */
export const joinTeamByCode = async (code: string): Promise<User> => {
    try {
        const response = await withToken(networkJoinTeamByCode, code)
        const { user } = response.data
        return user
    } catch (error) {
        return throwApiError(error, Constants.EDIT_USER_ERROR)
    }
}

/**
 * Method to get the current user's id based on the current token
 * @returns user's id of empty string
 */
export const getUserId = async (): Promise<string> => {
    const token = await EncryptedStorage.getItem('access_token')
    if (!token) {
        return ''
    }
    const decoded = jwt_decode(token) as any
    return decoded.sub
}
