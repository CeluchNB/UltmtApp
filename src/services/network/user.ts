import { API_URL_V1 } from 'react-native-dotenv'
import { ApiResponse } from '../../types/services'
import { CreateUserData } from '../../types/user'
import EncryptedStorage from 'react-native-encrypted-storage'
import axios, { AxiosResponse } from 'axios'

export const login = async (
    username: string,
    password: string,
): Promise<AxiosResponse> => {
    return await axios.post(`${API_URL_V1}/user/login`, {
        password,
        email: username,
    })
}

export const fetchProfile = async (token: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/user/me`, {
        headers: {
            Authorization: `Bearer ${token}`,
        },
    })
}

export const createAccount = async (
    profileData: CreateUserData,
): Promise<AxiosResponse> => {
    return await axios.post(`${API_URL_V1}/user`, {
        ...profileData,
    })
}

export const logout = async (token: string): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/user/logout`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const getLocalToken = async (): Promise<ApiResponse> => {
    try {
        const token = await EncryptedStorage.getItem('jwt_token')
        if (token) {
            return { data: token }
        }
    } catch (error) {
        throw error
    }
    throw new Error('No token available')
}

export const searchUsers = async (term: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/user/search?q=${term}`)
}

export const leaveTeam = async (
    token: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/user/leave/team?team=${teamId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const getPublicUser = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/user/${id}`)
}

export const leaveManagerRole = async (
    token: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/user/managerLeave?team=${teamId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const requestPasswordRecovery = async (
    email: string,
): Promise<AxiosResponse> => {
    return await axios.post(`${API_URL_V1}/user/requestPasswordRecovery`, {
        email,
    })
}

export const resetPassword = async (
    passcode: string,
    newPassword: string,
): Promise<AxiosResponse> => {
    return await axios.post(`${API_URL_V1}/user/resetPassword`, {
        passcode,
        newPassword,
    })
}

export const setOpenToRequests = async (
    token: string,
    open: boolean,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/user/open?open=${open}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const changeName = async (
    token: string,
    newFirstName: string,
    newLastName: string,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/user/changeName`,
        {
            newFirstName,
            newLastName,
        },
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const setPrivate = async (
    token: string,
    privateAccount: boolean,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/user/setPrivate?private=${privateAccount}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
            },
        },
    )
}

export const changeEmail = async (
    email: string,
    password: string,
    newEmail: string,
): Promise<AxiosResponse> => {
    return await axios.put(`${API_URL_V1}/user/changeEmail`, {
        email,
        password,
        newEmail,
    })
}
