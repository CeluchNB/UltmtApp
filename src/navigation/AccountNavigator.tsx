import * as React from 'react'
import { AccountStackParamList } from './../types/navigation'
import CreateAccountScreen from './../screens/CreateAccountScreen'
import CreateTeamScreen from './../screens/CreateTeamScreen'
import ForgotPasswordScreen from '../screens/ForgotPasswordScreen'
import LoginScreen from './../screens/LoginScreen'
import ManageTeamDetailsScreen from './../screens/ManageTeamDetailsScreen'
import ManageTeamsScreen from './../screens/ManageTeamsScreen'
import ProfileScreen from './../screens/ProfileScreen'
import PublicTeamScreen from './../screens/PublicTeamScreen'
import PublicUserScreen from './../screens/PublicUserScreen'
import RequestTeamScreen from './../screens/RequestTeamScreen'
import { RequestType } from './../types/request'
import RequestUserScreen from './../screens/RequestUserScreen'
import ResetPasswordScreen from '../screens/ResetPasswordScreen'
import RolloverTeamScreen from './../screens/RolloverTeamScreen'
import { createNativeStackNavigator } from '@react-navigation/native-stack'

const Stack = createNativeStackNavigator<AccountStackParamList>()

const AccountNavigator: React.FC<{}> = () => {
    return (
        <Stack.Navigator
            initialRouteName={'Login'}
            screenOptions={{ headerShown: false }}>
            <Stack.Screen name="Login" component={LoginScreen} />
            <Stack.Screen name="Profile" component={ProfileScreen} />
            <Stack.Screen
                name="CreateAccount"
                component={CreateAccountScreen}
            />
            <Stack.Screen name="CreateTeam" component={CreateTeamScreen} />
            <Stack.Screen name="ManageTeams" component={ManageTeamsScreen} />
            <Stack.Screen name="RequestTeam" component={RequestTeamScreen} />
            <Stack.Screen
                name="ManagedTeamDetails"
                component={ManageTeamDetailsScreen}
                initialParams={{ id: '', place: '', name: '' }}
            />
            <Stack.Screen
                name="PublicTeamDetails"
                component={PublicTeamScreen}
                initialParams={{ id: '', place: '', name: '' }}
            />
            <Stack.Screen
                name="RequestUser"
                component={RequestUserScreen}
                initialParams={{ type: RequestType.PLAYER }}
            />
            <Stack.Screen
                name="RolloverTeam"
                component={RolloverTeamScreen}
                initialParams={{ hasPendingRequests: false }}
            />
            <Stack.Screen
                name="PublicUserDetails"
                component={PublicUserScreen}
                initialParams={{
                    user: {
                        _id: '',
                        firstName: '',
                        lastName: '',
                        username: '',
                    },
                }}
            />
            <Stack.Screen
                name="ForgotPasswordScreen"
                component={ForgotPasswordScreen}
            />
            <Stack.Screen
                name="ResetPasswordScreen"
                component={ResetPasswordScreen}
            />
        </Stack.Navigator>
    )
}

export default AccountNavigator
