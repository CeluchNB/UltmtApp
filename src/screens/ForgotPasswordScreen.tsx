import * as React from 'react'
import { ForgotPasswordProps } from '../types/navigation'
import PrimaryButton from '../components/atoms/PrimaryButton'
import ScreenTitle from '../components/atoms/ScreenTitle'
import SecondaryButton from '../components/atoms/SecondaryButton'
import UserInput from '../components/atoms/UserInput'
import { getFormFieldRules } from '../utils/form-utils'
import { requestPasswordRecovery } from '../services/data/user'
import { size } from '../theme/fonts'
import validator from 'validator'
import { Controller, useForm } from 'react-hook-form'
import { StyleSheet, Text, View } from 'react-native'
import { useColors, useLazyData } from '../hooks'

const ForgotPasswordScreen: React.FC<ForgotPasswordProps> = ({
    navigation,
}) => {
    const { colors } = useColors()
    const [success, setSuccess] = React.useState(false)

    const {
        control,
        handleSubmit,
        reset,
        formState: { errors },
    } = useForm({
        defaultValues: {
            email: '',
        },
    })

    const { loading, fetch, error } = useLazyData(requestPasswordRecovery)

    const submitRequest = async (data: { email: string }) => {
        setSuccess(false)
        await fetch(data.email)
        reset({ email: '' })
        setSuccess(true)
    }

    const styles = StyleSheet.create({
        screen: {
            height: '100%',
            backgroundColor: colors.primary,
        },
        title: { alignSelf: 'center' },
        input: {
            width: '75%',
            alignSelf: 'center',
            marginTop: 20,
        },
        submitButton: {
            marginTop: 20,
        },
        button: {
            alignSelf: 'center',
        },
        error: {
            color: colors.error,
            width: '75%',
            alignSelf: 'center',
            marginTop: 2,
        },
        success: {
            color: colors.success,
            width: '75%',
            alignSelf: 'center',
            marginTop: 2,
            fontSize: size.fontMedium,
        },
    })

    return (
        <View style={styles.screen}>
            <ScreenTitle title="Forgot Password?" style={styles.title} />
            <Controller
                control={control}
                rules={getFormFieldRules('Email', true, undefined, undefined, [
                    {
                        test: (v: string) => {
                            return validator.isEmail(v)
                        },
                        message: 'Email must be in valid email format.',
                    },
                ])}
                render={({ field: { onChange, value } }) => {
                    return (
                        <UserInput
                            placeholder="Email"
                            onChangeText={onChange}
                            value={value}
                            style={styles.input}
                        />
                    )
                }}
                name="email"
            />
            {errors.email && (
                <Text style={styles.error}>{errors.email.message}</Text>
            )}
            {error && <Text style={styles.error}>{error.message}</Text>}
            {!error && success && (
                <Text style={styles.success}>
                    Check your inbox for a recovery code!
                </Text>
            )}
            <PrimaryButton
                text="Submit"
                onPress={handleSubmit(submitRequest)}
                loading={loading}
                style={[styles.button, styles.submitButton]}
            />
            <SecondaryButton
                text="I Have A Code"
                onPress={async () => {
                    navigation.navigate('ResetPasswordScreen')
                }}
                style={styles.button}
            />
        </View>
    )
}

export default ForgotPasswordScreen
