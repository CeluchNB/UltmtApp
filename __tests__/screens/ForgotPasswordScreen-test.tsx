import * as UserData from '../../src/services/data/user'
import { ForgotPasswordProps } from '../../src/types/navigation'
import ForgotPasswordScreen from '../../src/screens/ForgotPasswordScreen'
import { NavigationContainer } from '@react-navigation/native'
import React from 'react'
import renderer from 'react-test-renderer'
import { act, fireEvent, render } from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const reset = jest.fn()
const navigate = jest.fn()

const props: ForgotPasswordProps = {
    navigation: {
        navigate,
        reset,
    } as any,
    route: {} as any,
}

beforeEach(() => {
    reset.mockReset()
    navigate.mockReset()
})

describe('test forgot password screen', () => {
    it('should match snapshot', () => {
        const snapshot = renderer.create(
            <NavigationContainer>
                <ForgotPasswordScreen {...props} />
            </NavigationContainer>,
        )

        expect(snapshot).toMatchSnapshot()
    })

    it('test network error', async () => {
        const spy = jest
            .spyOn(UserData, 'requestPasswordRecovery')
            .mockClear()
            .mockReturnValueOnce(
                Promise.reject({
                    message: 'Unable to send email',
                }),
            )

        const { getByPlaceholderText, getByText, queryByText } = render(
            <NavigationContainer>
                <ForgotPasswordScreen {...props} />
            </NavigationContainer>,
        )

        fireEvent.changeText(getByPlaceholderText('Email'), 'test@email.com')
        fireEvent.press(getByText('Submit'))
        await act(async () => {})

        expect(queryByText('Unable to send email')).not.toBeNull()
        expect(spy).toHaveBeenCalled()
    })

    it('should call request password recovery with valid data', async () => {
        const spy = jest
            .spyOn(UserData, 'requestPasswordRecovery')
            .mockClear()
            .mockReturnValueOnce(Promise.resolve())

        const { getByPlaceholderText, getByText, queryByText } = render(
            <NavigationContainer>
                <ForgotPasswordScreen {...props} />
            </NavigationContainer>,
        )

        fireEvent.changeText(getByPlaceholderText('Email'), 'test@email.com')
        fireEvent.press(getByText('Submit'))
        await act(async () => {})

        expect(
            queryByText('Check your inbox for a recovery code!'),
        ).not.toBeNull()
        expect(spy).toHaveBeenCalledWith('test@email.com')
    })

    it('test form error', async () => {
        const spy = jest
            .spyOn(UserData, 'requestPasswordRecovery')
            .mockClear()
            .mockReturnValueOnce(Promise.resolve())

        const { getByPlaceholderText, getByText } = render(
            <NavigationContainer>
                <ForgotPasswordScreen {...props} />
            </NavigationContainer>,
        )

        fireEvent.changeText(getByPlaceholderText('Email'), 'testemailcom')
        fireEvent.press(getByText('Submit'))
        await act(async () => {})

        expect(spy).not.toHaveBeenCalled()
    })

    it('should navigate to reset password screen', async () => {
        const { getByText } = render(
            <NavigationContainer>
                <ForgotPasswordScreen {...props} />
            </NavigationContainer>,
        )

        fireEvent.press(getByText('I Have A Code'))
        expect(navigate).toHaveBeenLastCalledWith('ResetPassword')
    })
})
