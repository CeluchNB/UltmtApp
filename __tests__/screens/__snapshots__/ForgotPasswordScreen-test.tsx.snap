// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test forgot password screen should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#303030",
      "height": "100%",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "color": "#ffd600",
          "fontSize": 61.64383561643835,
          "fontWeight": "bold",
        },
        Object {
          "alignSelf": "center",
        },
      ]
    }
  >
    Forgot Password?
  </Text>
  <View
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 20,
        "placeholderTextColor": "#ffffff",
        "width": "75%",
      }
    }
  >
    <View
      pointerEvents="none"
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 6,
          },
          Object {
            "backgroundColor": "#303030",
            "borderColor": "#31adff",
            "borderRadius": 4,
            "borderWidth": 1,
          },
        ]
      }
      testID="text-input-outline"
    />
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
            },
            Object {
              "minHeight": 64,
              "paddingTop": 8,
            },
          ]
        }
      >
        <TextInput
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Email"
          placeholderTextColor="#ffffff"
          selectionColor="#ffd600"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "paddingHorizontal": 14,
                "zIndex": 1,
              },
              Object {
                "height": 56,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#ffd600",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="text-input-outlined"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#ffd600",
        "borderColor": "transparent",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 2,
        "marginTop": 20,
        "minWidth": "auto",
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0.75,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 1.5,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                Object {
                  "marginHorizontal": 8,
                },
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </View>
  <View
    collapsable={false}
    nativeID="animatedComponent"
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "transparent",
        "borderColor": "#ffd600",
        "borderRadius": 4,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "elevation": 0,
        "minWidth": 64,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 4,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#ffd600",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                undefined,
              ],
            ]
          }
        >
          I Have A Code
        </Text>
      </View>
    </View>
  </View>
</View>
`;
